# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


H₀: μ₀  = 123.7
H₁: μ > 123.7
правосторонній тест

In [1]:
import numpy as np
from scipy import stats

In [3]:
# УМОВА

time = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7   # середнє у магазині
alpha = 0.1
alpha2 = 0.05

In [ ]:
# ВИБІРКОВІ СТАТИСТИКИ

x_bar = np.mean(time)
s = np.std(time, ddof=1)    # стандартне відхилення з n-1
n = len(time)

print(f"x_bar = {x_bar:.3f}")
print(f"s = {s:.3f}")
print(f"n = {n:.3f}")

x_bar = 132.286
s = 10.981
n = 7.000


In [ ]:
# при alpha = 0.1

# t-СТАТИСТИКА
t_stat = (x_bar - mu0) / (s / np.sqrt(n))
print(f"t-статистика = {t_stat:.3f}")

# КРИТИЧНЕ ЗНАЧЕННЯ для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1 - alpha, df)
print(f"критичне значення = {t_crit:.3f}")

# p-VALUE
p_value = 1 - stats.t.cdf(t_stat, df)
print(f"р-значення = {p_value:.3f}")

t-статистика = 2.069
критичне значення = 1.440
р-значення = 0.042


In [18]:
# РІШЕННЯ при рівні значущості alpha = 0.1

if t_stat > t_crit:
    print(f"Відхиляємо H0: є підстави вважати, що середній час до покупки більший за 123.7 на рівні значущості {alpha}")
else:
    print(f"Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: є підстави вважати, що середній час до покупки більший за 123.7 на рівні значущості 0.1


ВИСНОВОК:

редизайн сайту справді подовжив час до покупки

In [11]:
# при рівні значущості alpha = 0.05

# КРИТИЧНЕ ЗНАЧЕННЯ для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1 - alpha2, df)
print(f"критичне значення = {t_crit:.3f}")

критичне значення = 1.943


In [19]:
# РІШЕННЯ при рівні значущості alpha = 0.05

if t_stat > t_crit:
    print(f"Відхиляємо H0: є підстави вважати, що середній час до покупки більший за 123.7 на рівні значущості {alpha2}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: є підстави вважати, що середній час до покупки більший за 123.7 на рівні значущості 0.05


Чи зміниться рішення при рівні значущості alpha = 0.05

ВИСНОВОК:

При рівні значущості alpha = 0.05 рішення не зміниться, час до покупки збільшився

In [23]:
# 2 СПОСІБ:

# scipy тест

res = stats.ttest_1samp(time, popmean=mu0, alternative="greater")
print(f"scipy: t = {res.statistic:.3f}, p-value = {res.pvalue:.5f}")



scipy: t = 2.069, p-value = 0.04202


In [22]:
res

TtestResult(statistic=np.float64(2.0687273144186475), pvalue=np.float64(0.042017654253682576), df=np.int64(6))

In [24]:
# РІШЕННЯ на основі р-значення

if res.pvalue < alpha:
    print(f"Відхиляємо H0: є підстави вважати, що середній час до покупки збільшився")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: є підстави вважати, що середній час до покупки збільшився


In [26]:
t_crit = stats.t.ppf(alpha, res.df)

if t_stat > t_crit:
    print(f"Відхиляємо H0: є підстави вважати, що середній час до покупки збільшився")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: є підстави вважати, що середній час до покупки збільшився


**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [27]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

=> тест для однієї пропорції

H₀: p₀  = 0.23  #впізнаваність логотипу не змінилась

H₁: p > 0.23   #впізнаваність логотипу зросла

правосторонній

In [36]:
# УМОВА

count = 311   # кількість "успіхів"
nobs = 1200    # обсяг вибірки
p0 = 0.23      # гіпотетична пропорція (H0)
alpha = 0.03

In [41]:
# Z-ТЕСТ для пропорції

stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha, method='normal')

print("тест для однієї пропорції")
print(f"Z-статистика = {stat:.4f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")

тест для однієї пропорції
Z-статистика = 2.3058
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


In [43]:
# РІШЕННЯ

if pval < alpha:
    print("Відхиляємо H₀")
    print("Є підстави вважати, що впізнаваність логотипу зросла")
else:
    print("Не відхиляємо H₀")
    print("Недостатньо підстав стверджувати, що впізнаваність логотипу зросла")

Відхиляємо H₀
Є підстави вважати, що впізнаваність логотипу зросла


ДОВІРЧИЙ ІНТЕРРВАЛ

97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127):

Нижня межа > 0.23 це додатково підтверджує висновок про зростання впізнаваності логотипу
на заданому рівні значущості alpha = 0.03 